autrainer/core/scripts/preprocess_script.py

```diff
@@ -196,39 +197,96 @@ def _process_chunk(
-            _create_subdirs(
-                os.path.join(dataset["path"], dataset["features_subdir"]),
-                unique_files,
+            data = autrainer.instantiate_shorthand(dataset)
+            loader = torch.utils.data.DataLoader(
+                torch.utils.data.ConcatDataset(
+                    data.train_dataset,
+                    data.dev_dataset,
+                    data.test_dataset
+                ),
+                shuffle=False,
+                num_workers=self.num_workers,
+                batch_size=1  #TODO: can we do it batched?
+            )
+            pipeline = SmartCompose(
+                [
+                    autrainer.instantiate_shorthand(t)
+                    for t in preprocess["pipeline"]
+                ]
             )
-            chunks = _split_chunks(unique_files, self.num_workers)
-            lock = tqdm.get_lock()
-            with tqdm(
-                total=len(unique_files),
-                desc=name,
-                disable=self.update_frequency == 0,
-            ) as pbar:
-                with ThreadPoolExecutor(self.num_workers) as executor:
-                    futures = [
-                        executor.submit(
-                            _process_chunk,
-                            chunk,
-                            dataset,
-                            preprocess,
-                            pbar,
-                            lock,
-                        )
-                        for chunk in chunks
-                    ]
-                    for future in futures:
-                        future.result()
+            for d, n in (
+                (data.train_dataset, "train"),
+                (data.dev_dataset, "dev"),
+                (data.test_dataset, "test"),
+            ):
+                loader = torch.utils.data.DataLoader(
+                    dataset=d,
+                    shuffle=False,
+                    num_workers=self.num_workers,
+                    batch_size=1  #TODO: can we do it batched?
+                )
+                for data in tqdm.tqdm(
+                    loader,
+                    total=len(loader),
+                    desc=f"{name}-{n}",
+                    disable=self.update_frequency == 0
+                ):
+                    #TODO: will be streamlined once we switch to dataclass
+                    index = d.df.index[data[2]]
+                    item_path = d.df.loc[index, d.index_column]
+                    out_path = Path(
+                        dataset["path"],
+                        features_subdir,
+                        os.path.basename(item_path),
+                    ).with_suffix("." + output_file_type)
+                    if os.path.exists(out_path):
+                        continue
+                    output_file_handler.save(
+                        out_path,
+                        pipeline(data[0], 0)  #TODO: why 0?
```

Comment:
The `0` is the index of the sample in the dataset, which we previously could not have known but needs to be passed here as some transforms may rely on it. Since preprocessing should be "static" it is hardcoded here for determinism, ref: https://github.com/autrainer/autrainer/blob/e1ecb6c2a4e98ed6e00abd9d6423117bfdb9abbb/autrainer/transforms/smart_compose.py#L93-L108

But yeah this would also be greatly streamlined once we switch to a dataclass for transforms and datasets.

Reply:
I thought so..So maybe I just leave it to 0 for now?

Reply:
Yeah i would say so, the dataclass refactor will probably change this part anyways.